### PR TITLE
feat: add image attachment support to chat

### PR DESCRIPTION
**Pull request by nakajima**
This PR adds support for sending E2E encrypted images in messages.

It uses the newly introduced `Attachment` and `RemoteAttachment` content types which you can read more about here: https://xmtp.org/blog/attachments-and-remote-attachments.

https://user-images.githubusercontent.com/483/227057859-b5dd6059-cdd5-4224-81ac-22d97029afe9.mov

I had to change some of the hardcoded height values that were in the messages area before to accommodate the composer growing when you're about to send an image. I also added a slight background color to the composer because it looked confusing to have a bunch of white space when you had an attachment preview in there. Feel free to change any of that if it doesn't work for you.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Enhancement (non-breaking small changes to existing functionality)

## How should this be tested?

You should be able to add an image by clicking the photo button in the chat view.

### CODE REVIEW

**Comment by bigint** [comment]
Can we make like a regular chat attachment to upload and load seamlessly? 🙇🏼 

We don't want to show attachment name, load button etc

Also is it possible to send text along with attachments?

**Comment by nakajima** [comment]
> Can we make like a regular chat attachment to upload and load seamlessly? 🙇🏼

So we don't recommend attachments get autoloaded by default, since it presents somewhat of a security risk. What do you think of adding the ability to automatically load attachments for "Following" conversations that are under a certain size, maybe 100MB?

> Also is it possible to send text along with attachments?

Not yet, but we can add this soon.

**Comment by bigint** [comment]
> So we don't recommend attachments get autoloaded by default, since it presents somewhat of a security risk. What do you think of adding the ability to automatically load attachments for "Following" conversations that are under a certain size, maybe 100MB?
> 
> >

yea good idea to add only for following users and we have have load button in requested profiles

Also we can explain something like this for users like twitter does instead just showing load button

image

**Comment by nakajima** [comment]
Cool, I'll get goin on that.

**Comment by bigint** [comment]
Amazing you are great @nakajima 💪👏

**Comment by nakajima** [comment]
Ok, just pushed https://github.com/lensterxyz/lenster/pull/2149/commits/528db214b4600b33347a3f9cc4ca4b1098c075e8. This makes it so that we auto load attachments for convos where you're following the other profile (unless the attachment is greater than 100mb).

I also introduced a new cache using indexed db to prevent loading the same attachments multiple times, since it can be a bit expensive.

Some screenshots:

|not following|too big|loader|
|-|-|-|
|Screenshot 2023-03-24 at 1 28 37 PM|Screenshot 2023-03-24 at 1 29 29 PM|Screenshot 2023-03-24 at 1 29 39 PM|

File: apps/web/src/components/Messages/Composer.tsx [hunk 1]
```diff
@@ -1,39 +1,107 @@
...
+  const canSendMessage = !disabledInput && !sending && (message.length > 0 || attachment);
 
   const handleSend = async () => {
     if (!canSendMessage) {
       return;
     }
     setSending(true);
-    const sent = await sendMessage(message);
+
+    var sent: boolean;
```
Comment (harish-sethuraman) [inline apps/web/src/components/Messages/Composer.tsx:69]:
```suggestion
    let sent: boolean;
```

File: apps/web/src/components/Messages/Composer.tsx [hunk 2]
```diff
@@ -1,39 +1,107 @@
...
+      const file = new File([encryptedEncodedContent.payload], 'XMTPEncryptedContent', {
+        type: attachment.mimeType
+      });
+
+      const lensterAttachment = await uploadFileToIPFS(file);
+      const cid = lensterAttachment?.item.replace('ipfs://', '');
+      const url = `https://${cid}.ipfs.w3s.link`;
+
+      const remoteAttachment: RemoteAttachment = {
+        url: url,
```
Comment (harish-sethuraman) [inline apps/web/src/components/Messages/Composer.tsx:85]:
```suggestion
        url,
```

File: apps/web/src/components/Messages/Composer.tsx [hunk 3]
```diff
@@ -58,27 +126,77 @@ const Composer: FC<ComposerProps> = ({ sendMessage, conversationKey, disabledInp
...
+      fileReader.readAsArrayBuffer(file);
+    } else {
+      setAttachment(null);
+    }
+  };
+
+  function onDismiss() {
+    setAttachment(null);
+
+    const el = fileInputRef?.current;
```
Comment (harish-sethuraman) [inline apps/web/src/components/Messages/Composer.tsx:159]:
```suggestion
    const el = fileInputRef.current;
```
all reference will always have `current`?

File: apps/web/src/components/Messages/MessageContent.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+type MessageContentProps = {
+  message: DecodedMessage;
+  profile: Profile | undefined;
+  sentByMe: boolean;
+};
+
+const MessageContent = ({ message, profile, sentByMe }: MessageContentProps): JSX.Element => {
+  function content(): JSX.Element {
+    if (message.error) {
+      return <span>{`Error: ${message.error}`}</span>;
```
Comment (harish-sethuraman) [inline apps/web/src/components/Messages/MessageContent.tsx:17]:
```suggestion
      return <span>Error: {message.error}</span>;
```

**Comment by bigint** [comment]
@nakajima Loader state will cause CLS, can we keep the same width as warn state for loading card 🙇🏼 ?

**Comment by bigint** [comment]
Also build failed 🙇🏼 

image

**Comment by nakajima** [comment]
Pushed a fix for that build error!

> Loader state will cause CLS, can we keep the same width as warn state for loading card

The loader state only appears when you click "View" on an unloaded attachment, so I don't think there should really be more than one on the screen at once. It probably doesn't represent a big CLS issue in that case I think?

**Comment by bigint** [comment]
@nakajima yea, it will still be CLS even after clicking. Here is an example video

Consider we might have some kind of reaction button next to the warning card it will get shifted on clicking view button and there is possibility to react for the message.

[layout-instability2.webm](https://user-images.githubusercontent.com/69431456/228016146-0bb8feb9-71c3-4cf3-b30b-39cf49f41e4e.webm)

**Comment by bigint** [comment]
@nakajima this one

image

**Comment by bigint** [comment]
@nakajima looks cool, we can use small spinner for this

**Comment by nakajima** [comment]
Screenshot 2023-03-28 at 8 16 43 AM

**Comment by nakajima** [comment]
How does that look ^^

**Comment by bigint** [comment]
@nakajima Reduce the padding-y a bit 🙇🏼

**Comment by nakajima** [comment]
Screenshot 2023-03-28 at 8 22 40 AM

^^

**Comment by bigint** [comment]
Perfect

File: apps/web/src/store/attachment.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import { Localstorage } from 'data/storage';
+import { del, get, set } from 'idb-keyval';
```
[the remark this answers is not in the document]
Comment (yash-luna) [reply apps/web/src/store/attachment.ts:2]: Like @nakajima mentions, storing too many images is not going to be a problem right away

@bigint do you think we can push this update to gauge adoption and then follow up with a PR to evict old data?

**Comment by bigint** [comment]
@nakajima It would be great if you fix the conflicts, I'll merge after that 🚀 🙇🏼

**Review by bigint — APPROVED** [review]
**LGTM** :shipit: 

Rad LGTM

**Review by saagu335 — APPROVED** [review]
@nakajima Please fix the conflicts and `@` me I'll merge this ASAP 🙇 

PS: we migrated to pnpm from yarn

**Comment by bigint** [comment]
Gonna close this for now!

We need to fix the speed issue first, later we can ship more features 🙇🏼 

RN it takes 40s to load my inbox 😢

**Comment by bigint** [comment]
@nakajima It would be great if you resolve the conflicts 🙇🏼

**Comment by daria-github** [comment]
@bigint conflicts should be resolved in this PR: https://github.com/lensterxyz/lenster/pull/2845

**Comment by bigint** [comment]
Closing this and continued here #2845

**Pull request closed without merging**